src/org/opensolaris/opengrok/authorization/AuthorizationFramework.java

```diff
@@ -683,31 +762,52 @@ private boolean checkAll(HttpServletRequest request, String cache, Nameable enti
 
         stats.addRequest("authorization_cache_misses");
 
-        long time = System.currentTimeMillis();
-
-        boolean overallDecision = performCheck(entity, pluginPredicate, skippingPredicate);
+        long time = 0;
+        boolean overallDecision = false;
+        
+        lock.readLock().lock();
+        try {
+            // Make sure there is a HTTP session that corresponds to current plugin version.
+            HttpSession session;
+            if (((session = request.getSession(false)) != null) && isSessionInvalid(session)) {
+                session.invalidate();
+                stats.addRequest("authorization_sessions_invalidated");
+            }
+            request.getSession().setAttribute(SESSION_VERSION, getPluginVersion());
```

Comment:
does this need to be locked under the lock if this does not interfere with the stack?

Reply:
ok, clear enough